csharp/ql/src/semmle/code/csharp/exprs/Expr.qll

```diff
@@ -447,6 +453,19 @@ class PositionalPatternExpr extends Expr, @positional_pattern_expr {
   override string getAPrimaryQlClass() { result = "PositionalPatternExpr" }
 }
 
+/** A unary pattern. For example, `not 1`. */
+abstract class UnaryPatternExpr extends PatternExpr, @unary_pattern_expr {
```

Comment:
I see that `PatternExpr`, `Expr`, ... are also not abstract, so it makes sense to follow that pattern. But out of curiosity, wouldn't it make sense to make `UnaryPatternExpr` abstract? There can't be any direct instance of it, only inherited ones.

Reply:
We should never use `abstract` in public classes because sub-classing (as in, define a subset of the class) becomes very cumbersome. When a class `A` extends an abstract class `B` it means that `A` _adds_ to the class `B` (as opposed to defining a subset of `B`).